README.md
# Polyline_do
## Mapbox supported

Polyline encoding is a lossy compression algorithm that allows you to store a series of coordinates as a single string.

Polyline is a dart port of Google's Polyline Algorithm explained [here](https://developers.google.com/maps/documentation/utilities/polylinealgorithm). Inspired by Mapbox's polyline.js. Compatible with Dart 2.

# Usage 

### A simple usage example:
Two named constructors are provided Polyline.Decode and Polyline.Encode.

Calling Polyline.Encode will compute the __encodedString__ and set the passed in coordinates to __decodedCoords__ on the instance of Polyline so that each instance of Polyline has access to the correct encoded string and subsequent decoded coordinates. Likewise, calling Polyline.Decode computes the list of coordinates from the encoded string and sets the passed in encoded string to the instance of Polyline.

Note that a precision of 6 is standard.

import 'package:polyline_do/polyline_do.dart';

```
import 'package:polyline_do/polyline_do.dart';

void main() {
  Polyline polyline;

  /// List<List<double> coordinates;
  const coordinates = [
    [33.80119, -84.34788],
    [35.10566, -80.8762],
    [30.4526, -81.71116],
    [28.57888, -81.2717]
  ];
  const precision = 5;
  const encoded = 'mxhmEfeyaO}w}F_aeTrxk[nabDv}lJsytA';

  // Encode a list of coordinates with precision 5 to produce the encoded string
  polyline = Polyline.Encode(decodedCoords: coordinates, precision: 5);
  print('Encoded String: ${polyline.encodedString}, Coords: ${polyline.decodedCoords}');

  // Decode an encoded string to a list of coordinates
  polyline = Polyline.Decode(encodedString: encoded, precision: precision);
  print('Decoded Coords: ${polyline.decodedCoords}');
  print('String: ${polyline.encodedString}');

  // Calculate the distance of an encoded polyline, and decode the polyline
  polyline =  Polyline.Distance(encodedString: encoded,  unit: 'kilometers');
  // By calling  length the encodedString, decodedCoords, and distance variables
  // of the Polyline class are available
  print('Distance: ${polyline.distance!.floor()}km , Encoded String: ${polyline.encodedString} Decoded Coords: ${polyline.decodedCoords}');

}

  ```

  

CHANGELOG.md
## 0.1.0

* Null-safety migration


test/polyline_do_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:polyline_do/polyline_do.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/polyline_do.dart
library polyline_do;

import 'dart:math' show pow, sin, cos, pi, sqrt;

/*
* Polyline Encoding Algorithm
*
* - Decode encoded polyline
* - Encode polyline
*
* Todo add merge polylines
* Todo add merge multiple polylines
* Todo add from geoJson
* Todo add haversine
* Todo add haversine distance
* Todo write tests
* */


class Polyline {
  //  --- Instance Variables
  bool get isPolyline => true;
  late String encodedString;
  List<List<double>> decodedCoords = [];
  int precision = 6;
  double? distance;
  String unit = 'kilometers';

  //  --- Named Constructors
  /// .Decode(str, p)
  Polyline.Decode({ required this.encodedString, required this.precision  }) {
    decodedCoords = _decode(encodedString, precision);
    encodedString = encodedString;
  }

  /// .Encode(coords, p)
  Polyline.Encode({required this.decodedCoords, required this.precision}) {
    encodedString = _encode_poly(decodedCoords, precision);
    decodedCoords = decodedCoords;
  }

  // .Distance(str, u)
  Polyline.Distance({required this.encodedString, required this.unit}) {
    distance = _length(encodedString, unit);   // _length sets decodedCoords
    encodedString = encodedString;
  }




  //  ------------- ENCODE & DECODE -------------

  /// Decodes encoded polyline string to a [latitude, longitude] coordinates list.
  /// @decode_poly Function
  /// @param {String} str
  /// @param {int} precision
  /// @returns {List<dynamic>}
  List<List<double>> _decode(String str, int precision) {
    int index = 0,
        lat = 0,
        lng = 0,
        shift = 0,
        result = 0,
        latitude_change,
        longitude_change;
        
    int? byte;
    num factor = pow(10, precision is int ? precision : 5);
    
    // ignore: omit_local_variable_types
    List<List<double>> coordinates = [];

    // Coordinates have variable length when encoded, so just keep
    // track of whether we've hit the end of the string. In each
    // loop iteration, a single coordinate is decoded.
    while (index < str.length) {
      // Reset shift, result, and byte
      byte = null;
      shift = 0;
      result = 0;

      do {
        byte = str.codeUnitAt(index++) - 63;
        result |= (byte & 0x1f) << shift;
        shift += 5;
      } while (byte >= 0x20);

      latitude_change = (((result & 1) == 1) ? ~(result >> 1) : (result >> 1));

      shift = result = 0;

      do {
        byte = str.codeUnitAt(index++) - 63;
        result |= (byte & 0x1f) << shift;
        shift += 5;
      } while (byte >= 0x20);

      longitude_change = ((result & 1) == 1 ? ~(result >> 1) : (result >> 1));

      lat += latitude_change;
      lng += longitude_change;

      coordinates.add([lat / factor, lng / factor]);
    }

    decodedCoords = coordinates;
    return coordinates;
  }

  /// Encodes the given [latitude, longitude] coordinates list to an encoded string.
  /// @encode_poly Function
  /// @param {List<dynamic>} coordinates
  /// @param {int} precision
  /// @returns {String}
  String _encode_poly(List<List<double>> coordinates, int precision) {
    if (coordinates.length == null) {
      return '';
    }

    num factor = pow(10, precision is int ? precision : 5);

    var output = _encode(coordinates[0][0], 0, factor.toInt()) +
        _encode(coordinates[0][1], 0, factor.toInt());

    for (var i = 1; i < coordinates.length; i++) {
      var a = coordinates[i], b = coordinates[i - 1];
      output += _encode(a[0], b[0], factor.toInt());
      output += _encode(a[1], b[1], factor.toInt());
    }

    return output;
  }

  /// Returns the character string
  /// @param {double} current
  /// @param {double} previous
  /// @param {int} factor
  /// @returns {String}
  String _encode(double current, double previous, int factor) {
    final _current = (current * factor).round();
    final _previous = (previous * factor).round();

    var coordinate = _current - _previous;
    coordinate <<= 1;
    if (_current - _previous < 0) {
      coordinate = ~coordinate;
    }

    var output = '';
    while (coordinate >= 0x20) {
      output += String.fromCharCode((0x20 | (coordinate & 0x1f)) + 63);
      coordinate >>= 5;
    }
    output += String.fromCharCode(coordinate + 63);
    return output;
  }


  /// Calculate the distance of the polyline. If radius is not provided, distance is flat, else distance is haversine distance
  /// NOTE: Support flat surface and sphere
  /// @param {string} polyline - The polyline to calculate from
  /// @param  String of 'meter' or 'kilometer' unit - returned unit value.
  /// @return {double} length - unit based on options.radius unit
  double _length(String polyline, String unit) {
    List<List<double>> decodedPolyline;
    decodedPolyline = _decode(polyline, 5);

    // setting class instance vars
    decodedCoords  = decodedPolyline;

    double distanceOfDecoded = 0;
    for (var i = 0; i < decodedPolyline.length - 1; i++) {
      final lat = [decodedPolyline[i][0], decodedPolyline[i][1]];
      final lon = [decodedPolyline[i + 1][0], decodedPolyline[i + 1][1]];

      distanceOfDecoded += _haversineDistance(lat, lon);
    }

    if(unit == 'meter') return  distanceOfDecoded * 1000;
    if(unit == 'kilometer') return distanceOfDecoded;

    distance = distanceOfDecoded;
    return distanceOfDecoded;
  }

  //  ------------- HAVERSINE -------------

  /// Convert Degree to Radian
  /// @param {double} deg - Degree
  /// @return {double} Radia
  double _degToRad(double deg) {
    return (deg * pi) / 180;
  }


  /// Calculate haversine of a number
  /// @param {double} number - input number
  /// @return {double} haversine
  double _haversine(double number) {
    if(number == null) {
      throw NullThrownError;
    }
    return pow(sin(number / 2), 2).toDouble();
  }


   /// Calculate the haversine distance between 2 points
   /// on the Earth, using radius of 6371 km
   /// @param {List<double>} point1 - lat, lon are mandatory
   /// @param {List<double>} point2 - lat, lon are mandatory
   /// @return {double} distance
  double _haversineDistance(List<double> _point1, List<double> _point2) {
    _point1.map((item) => item != null ?  _degToRad(item) : throw NullThrownError);
    _point2.map((item) => item != null ?  _degToRad(item) : throw NullThrownError);

    const radius = 6371;
    final point1 = [_point1[0], _point1[1] ];
    final point2 = [_point2[0], _point2[1] ];

    final a = _haversine(point2[0] - point1[0]);
    final b = cos(point1[0]) * cos(point2[0]) * _haversine(point2[1] - point1[1]);
    final distance = 2 * radius * sqrt(a + b);

    return distance;
  }
}


example/polyline_example.dart
import 'package:polyline_do/polyline_do.dart';

void main() {
  Polyline polyline;

  /// List<List<double> coordinates;
  const coordinates = [
    [33.80119, -84.34788],
    [35.10566, -80.8762],
    [30.4526, -81.71116],
    [28.57888, -81.2717]
  ];
  const precision = 5;
  const encoded = 'mxhmEfeyaO}w}F_aeTrxk[nabDv}lJsytA';

  // Encode a list of coordinates with precision 5 to produce the encoded string
  polyline = Polyline.Encode(decodedCoords: coordinates, precision: 5);
  print('Encoded String: ${polyline.encodedString}, Coords: ${polyline.decodedCoords}');

  // Decode an encoded string to a list of coordinates
  polyline = Polyline.Decode(encodedString: encoded, precision: precision);
  print('Decoded Coords: ${polyline.decodedCoords}');
  print('String: ${polyline.encodedString}');

  // Calculate the distance of an encoded polyline, and decode the polyline
  polyline =  Polyline.Distance(encodedString: encoded,  unit: 'kilometers');
  // By calling  length the encodedString, decodedCoords, and distance variables
  // of the Polyline class are available
  print('Distance: ${polyline.distance!.floor()}km , Encoded String: ${polyline.encodedString} Decoded Coords: ${polyline.decodedCoords}');

}